custom/icds_reports/sqldata/exports/poshan_progress_report.py

```diff
@@ -0,0 +1,188 @@
+            self.report_type = 'quarter'
+            self.quarter = self.config['quarter']
+            self.year = self.config['year']
+
+    def _generate_quarter_years(self):
+        months = []
+        end_month = self.quarter * 3
+        for i in range(end_month - 2, end_month + 1):
+            months.append(date(self.year, i, 1))
+        return months
+
+    def structure_data(self, headers, cols, rows):
+        """
+        :param headers: headers of the excel file
+        :param cols: columns mapped with headers
+        :param rows: row data
+        :return: list of list contain the excel rows
+        """
+        excel_rows = [headers]
+        for row in rows:
+            row_data = []
+            for col in cols:
+                row_data.append(row[col])
+            excel_rows.append(row_data)
+        return excel_rows
+
+    def quarter_wise(self, filters, order_by):
+        """
+        :param filters: quaterwise filter (months in [month array])
+        :param order_by: order by columns
+        :return: excel_rows
+        """
+        query_set = PoshanProgressReportView.objects.filter(**filters).order_by(*order_by)
+        cols = COLS_COMPREHENSIVE[:]
+        cols.append('district_id')  # used as key for the dict
+        data = query_set.values(*cols)
+        row_data_dict = {}
+        dummy_row = ['', ''] + [0 for _ in range(0, len(COLS_COMPREHENSIVE) - 2)]
+        headers = HEADERS_COMPREHENSIVE
+        # generate a dict
+        # {'disitrict_id': [contains the excel row with sum of col values for all months eg. m1+m2+m3]}
+        for row in data:
+            row_data = dummy_row
+            for k, v in row.items():
+                if k == 'state_name' or k == 'district_name':
+                    row_data[COLS_COMPREHENSIVE.index(k)] = v
+                elif k != 'district_id':
+                    row_data[COLS_COMPREHENSIVE.index(k)] += v
+            row_data_dict[row['district_id']] = row_data
+
+        # Calculating Average
+        # m1+m2+m3/3
+        for k, v in row_data_dict.items():
+            for col in COLS_COMPREHENSIVE[2:]:
+                if col != 'state_name' and col != 'district_name':
+                    row_data_dict[k][COLS_COMPREHENSIVE.index(col)] = v[COLS_COMPREHENSIVE.index(col)] / 3
+
+        # Calculating Percentage
+        # percent(current col) = 100 * (actual_value(prev col) / expected_value(prev prev col))
+        for k, v in row_data_dict.items():
+            i = 8
+            while i < len(COLS_COMPREHENSIVE):
+                row_data_dict[k][i] = 100 * (v[i - 1] / v[i - 2]) if v[i - 2] else 0
+                i = i + 3
+
+        if self.layout != 'comprehensive':
+            headers = HEADERS_SUMMARY
+            cols = COLS_SUMMARY[:]
+            extra_columns = list(set(COLS_COMPREHENSIVE) - set(COLS_SUMMARY))
+            indexes_to_remove = [COLS_COMPREHENSIVE.index(col) for col in extra_columns]
+            for k, v in row_data_dict.items():
```

Comment:
do you think just traversing through `COLS_SUMMARY` and pulling out just the value of i'th element from v would have been easier than finding out the difference between two set and excluding them later?
```
for k, v in row_data_dict.items():
    row_data_dict[k] = [v[COLS_COMPREHENSIVE.index(column)] for column in cols ]
```

Reply:
updated